finetune/util/metrics.py

```diff
@@ -120,61 +121,122 @@ def sequence_labeling_token_counts(true, predicted):
     return d
 
 
-def seq_recall(true, predicted, count_fn):
+def calc_recall(TP, FN):
+    try:
+        return TP / float(FN + TP)
+    except ZeroDivisionError:
+        return 0.0
+
+
+def calc_precision(TP, FP):
+    try:
+        return TP / float(FP + TP)
+    except ZeroDivisionError:
+        return 0.0
+
+
+def calc_f1(recall, precision):
+    try:
+        return 2 * (recall * precision) / (recall + precision)
+    except ZeroDivisionError:
+        return 0.0
+
+
+def seq_recall(true, predicted, span_type="token"):
+    count_fn = get_seq_count_fn(span_type)
     class_counts = count_fn(true, predicted)
     results = {}
     for cls_, counts in class_counts.items():
         FN = len(counts['false_negatives'])
         TP = len(counts['correct'])
-        try:
-            results[cls_] = TP / float(FN + TP)
-        except ZeroDivisionError: 
-            results[cls_] = 0.
+        results[cls_] = calc_recall(TP, FN)
     return results
 
 
-def seq_precision(true, predicted, count_fn):
+def seq_precision(true, predicted, span_type="token"):
+    count_fn = get_seq_count_fn(span_type)
     class_counts = count_fn(true, predicted)
     results = {}
     for cls_, counts in class_counts.items():
         FP = len(counts['false_positives'])
         TP = len(counts['correct'])
-        try:
-            results[cls_] = TP / float(FP + TP)
-        except ZeroDivisionError:
-            results[cls_] = 0.
+        results[cls_] = calc_precision(TP, FP)
     return results
 
-def micro_f1(true, predicted, count_fn):
+
+def micro_f1(true, predicted, span_type="token"):
+    count_fn = get_seq_count_fn(span_type)
     class_counts = count_fn(true, predicted)
     TP, FP, FN = 0, 0, 0
     for cls_, counts in class_counts.items():
         FN += len(counts['false_negatives'])
         TP += len(counts['correct'])
         FP += len(counts['false_positives'])
-    recall = TP/float(FN + TP)
-    precision = TP / float(FP + TP)
-    f1 = 2 * (recall * precision) / (recall + precision)
-    return f1
+    recall = calc_recall(TP, FN)
+    precision = calc_precision(TP, FP)
+    return calc_f1(recall, precision)
+
+
+def per_class_f1(true, predicted, span_type="token"):
+    """
+    F1-scores per class
+    """
+    count_fn = get_seq_count_fn(span_type)
+    class_counts = count_fn(true, predicted)
+    results = {}
+    for cls_, counts in class_counts.items():
+        results[cls_] = {}
+        FP = len(counts["false_positives"])
+        FN = len(counts["false_negatives"])
+        TP = len(counts["correct"])
+        recall = calc_recall(TP, FN)
+        precision = calc_precision(TP, FP)
+        results[cls_]["support"] = FN + TP
+        results[cls_]["f1-score"] = calc_f1(recall, precision)
+    return results
+
+
+def model_f1(true, predicted, span_type="token", average=None):
+    """
+    If average = None, return per-class F1 scores
+    """
+    if average == "micro":
```

Comment:
The flow of this function is kind of confusing:

I think you could do something like:

```python3
if average == "micro":
    return micro_f1(true, predicted, span_type)

f1_by_class = per_class_f1(true, predicted, span_type)
f1s = [f1s[cls_]['f1-score'] for cls_ in f1_by_class]
supports = [f1s[cls_]['support'] for cls_ in f1_by_class]

if average == 'weighted':
    return np.average(np.average(f1s, weights=np.array(supports))
if average == "macro":
    return np.average(f1s)
else:
    return f1s
```

Reply:
cleaned up.